Upstream information

v-mas opened the pull request:
I have added:
* display of upstream status in front of branch name (with option to hide it)
* fetch option (fetches changes from tracked branch to given branch)

also i have changed to use vscode provided api for gathering git information where possible (if it is not ok to use that, i can change it) - benefit of that is automatic refresh of VCS sidebar

I have tested it only on VSCode version 1.79.0, and git version 2.25.1

v-mas committed:
Display upstream status beside branch name

v-mas committed:
Added missing semicolons, extract custom git command execution to method

v-mas committed:
Add ability to fetch branch changes

m-hall commented on package.json:
```diff
@@ -139,12 +153,20 @@
                 {
-                    "command": "scm-local-branches.branch.delete-upstream",
+                    "command": "scm-local-branches.branch.unset-upstream",
                     "when": "view == scm-local-branches && viewItem == activeBranch"
```
Add a condition to also check if the branch currently has an upstream.

Alternatively, add a check in the unset upstream command itself to handle the case gracefully.
Right now it sends an error when there is no upstream.

m-hall commented on package.json:
```diff
@@ -26,6 +26,12 @@
+                    "default": true,
+                    "markdownDescription": "Show upstream information beside branch name",
+                    "scope": "window"
+                },
```
Change the setting to be `Show Upstream Status` instead of "information".
I was expecting this to show the name of the upstream rather than the sync status based on the title and description.

Could be worth adding a separate setting for `Show upstream Name`, if that's also desired.

v-mas replied:
in most cases the upstream branch name will be the same as name of local branch, so I have added that information to tooltip (it is shown on hover on menu item)
also the name of the branch can be long, and displaying it on sidebar would make it less readable

m-hall commented on src/git.ts:
```diff
@@ -122,159 +122,123 @@ export class Git implements vscode.Disposable {
+    }
 
+    public async unsetUpstream(branch: Branch): Promise<void> {
         try {
```
Mentioned above, but either guard against the upstream already not existing, or prevent this command from being available when it doesn't have one.

v-mas replied:
check added to package.json

m-hall commented on src/git.ts:
```diff
@@ -122,159 +122,123 @@ export class Git implements vscode.Disposable {
+        const gitBranches = await repo.getBranches({ remote: false })
+            .then<GitBranch[]>((all) => Promise.all(all.map((b) => repo.getBranch(b.name!!))));
+        const head = repo.state.HEAD;
+        const viewBranches: Branch[] = gitBranches.map((branch) => {
```
Previously the branches would be sorted alphabetically by git cmd. The Git extension doesn't seem to keep them sorted.
I'm ok if you want to add a setting to change the sorting method (unsorted/alphabetic/date changed/etc), but at a minimum it would be best if the sorting method stays the same for now.

v-mas replied:
in my test I had always got branches in name order. I had added the sorting so let me know if it does not work for you

m-hall replied:
Branches are sorted correctly for me again. 👍

m-hall reviewed (changes requested):
Branch list is no longer sorted alphabetically.

Generally I like what you're doing with this change. I love that it responds much faster to changes now <3

m-hall commented on package.json:
```diff
@@ -139,12 +153,20 @@
+                    "command": "scm-local-branches.branch.fetch",
+                    "when": "view == scm-local-branches && viewItem == branch"
+                },
+                {
```
Does checkout via the git extension not already fetch? It seems like it's usually up to date for me already.

v-mas replied:
I've checked: checkout does only that. It does not fetch changes nor pull

m-hall commented:
A thought also occurs:
Would probably make sense to either add a fetch all action, or modify the refresh action to perform a fetch all on the repo to make sure things are up to date.

v-mas committed:
Add branch/upstream info to branch item hover tooltip
sort branches by name
distinguish between branches that have upstream and one that do not

v-mas commented:
> A thought also occurs: Would probably make sense to either add a fetch all action, or modify the refresh action to perform a fetch all on the repo to make sure things are up to date.

I'm more willing adding separate action to "fetch all branches" - something like double arrow down, or download icon (cloud-download-outline from ionic icon pack)
the only problem is when local branch is ahead and behind at the same time - in such case branch will stay as is

m-hall commented on src/branch-tree-provider.ts:
```diff
@@ -45,20 +45,33 @@ export class BranchTreeProvider implements vscode.TreeDataProvider<Branch>, vsco
             }
             
             const item = new vscode.TreeItem(displayName);
+            item.tooltip = `name: ${element.branchName}`;
```
Nice touch! I like the tooltip :)

m-hall commented on package.json:
```diff
@@ -138,35 +138,23 @@
                 },
```
The option to delete a (non-active) branch that has an upstream should still be available.

If you want, you could modify the name to be "delete local branch" or add a confirmation if the branch has an upstream.
I don't consider either of those extras necessary, but I do think delete should still be available.

v-mas replied:
I just missed it, thanks for pointing out

m-hall reviewed (approved):
Approved provided the delete option is fixed. :)
Thanks for contributing!

m-hall commented:
> I'm more willing adding separate action to "fetch all branches" - something like double arrow down, or download icon (cloud-download-outline from ionic icon pack) the only problem is when local branch is ahead and behind at the same time - in such case branch will stay as is

Ahh ok, when I think of fetching, it's getting new branch meta from upstream, but not actually pushing or pulling (you can tie them together in your git config I believe, and I have not done that).
So your fetch options are more like "Sync changes with upstream" rather than "Fetch branch status from upstream".

Mild difference in semantics.

I agree I wouldn't really want a "sync all branches" in that case.

Might be worth revisiting the naming for consistency, since VSCode seems to use "sync" for those operations.

v-mas committed:
Renamed from fetch to sync; Added order to context actions

Regenhardt commented:
If I understand correctly, executing "sync" now does a git fetch (load upstream metadata without pulling)? I thought in vscode sync means pull+push?

v-mas commented:
I don't really know how the sync works in vscode, the one added does fast-forward 'git fetch' without checking out the branch.  It may be other usecase than the one that is directly in vscode, as their sync is available only for active branch.

m-hall commented:
Technically speaking, it's a `fetch`, which may be configured to `pull` as well.

There may have been some confusion with it, as it was behaving as a `pull` on my system, which is why I wanted it to be called `sync`.

VSCode's `sync` does a `pull+push`.

I'll probably go in and iron it out next week when I've got some free time.
